File: lib/services/user_service.dart
import 'package:logging/logging.dart';
import 'package:mongo_dart/mongo_dart.dart';

import '../models/user/user.dart';

class UserService {
  UserService(Db db) : _coll = db.collection('users');

  final _logger = Logger('NOMAC | UserService');
  final DbCollection _coll;

  Future<NomacUser?> getUserByDiscordId(String discordId) async {
    final result = await _coll.findOne(where.eq('discord_id', discordId));

    if (result == null) {
      return null;
    }

    return NomacUser.fromJson(result);
  }

  Future<bool> createUser(NomacUser user) async {
    try {
      await _coll.insert(user.toJson());
      return true;
    } catch (err) {
      _logger.severe(err);
      return false;
    }
  }

  Future<NomacUser> updateUser(NomacUser user) async {
    var result = await _coll.findOne(where.eq('discord_id', user.discordId));

    if (result == null) {
      await createUser(user);
      result = await _coll.findOne(where.eq('discord_id', user.discordId));
    }

    user.toJson().entries.forEach((element) {
      result![element.key] = element.value;
    });

    await _coll.replaceOne(where.eq('_id', result!['_id']), result);
    return user;
  }

  Future<bool> deleteUser(NomacUser user) async {
    try {
      await _coll.remove(where.eq('discord_id', user.discordId));
      return true;
    } catch (err) {
      _logger.severe(err);
      return false;
    }
  }
}

File: src/services/user_service.ts
import { Collection, Db } from "mongodb";
import { autoInjectable } from "tsyringe";
import { IUser } from "../types/user_types";

@autoInjectable()
export default class UserService {
	private coll: Collection;

	constructor(private db: Db) {
		this.coll = db.collection("users");
	}

	getUserByDiscordId = async (discord_id: string): Promise<IUser> => {
		return this.coll.findOne({
			discord_id,
		}) as Promise<IUser>;
	};

	updateUserByDiscordId = async (discord_id: string, data: Partial<Omit<IUser, "discord_id">>) => {
		return this.coll.findOneAndUpdate({ discord_id }, { $set: data });
	};

	deleteUser = async (discord_id: string) => {
		return this.coll.findOneAndDelete({ discord_id });
	};
}
